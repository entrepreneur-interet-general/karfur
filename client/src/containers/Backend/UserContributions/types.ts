import { ObjectId } from "mongodb";

export interface FormattedUserContribution {
  titreInformatif: string;
  titreMarque?: string;
  typeContenu: "dispositif" | "demarche";
  nbMercis: number;
  nbVues: number;
  responsabilite: string | null;
  _id: ObjectId;
  status: string;
  mainSponsor?: string | null;
  isAuthorizedToDelete: boolean;
}
