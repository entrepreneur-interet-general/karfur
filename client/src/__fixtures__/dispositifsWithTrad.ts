const content1 = {
  avancementExpert: 0,
  avancementTrad: 0.7333333333333333,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "Demander un logement social (HLM)",
  tradStatus: "À revoir",
  typeContenu: "demarche",
  _id: "id1",
};

const content2 = {
  avancementExpert: 0.5,
  avancementTrad: 1,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "T2",
  titreMarque: "TM2",
  tradStatus: "En attente",
  typeContenu: "dispositif",
  _id: "id2",
};
const content3 = {
  avancementExpert: 0,
  avancementTrad: 0,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "T3",
  titreMarque: "TM3",
  tradStatus: "À traduire",
  typeContenu: "dispositif",
  _id: "id3",
};

const content4 = {
  avancementExpert: 0,
  avancementTrad: 1,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "T4",
  titreMarque: "TM4",
  tradStatus: "Validée",
  typeContenu: "demarche",
  _id: "id4",
};

const content5 = {
  avancementExpert: 0,
  avancementTrad: 1,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "T5",
  titreMarque: "TM5",
  tradStatus: "À traduire",
  typeContenu: "demarche",
  _id: "id5",
};

const content6 = {
  avancementExpert: 0,
  avancementTrad: 0.7333333333333333,
  created_at: "2019-11-08T10:53:52.398Z",
  lastTradUpdatedAt: 1612514769665,
  nbMots: 812,
  titreInformatif: "TI6",
  titreMarque: "TM6",
  tradStatus: "À revoir",
  typeContenu: "dispositif",
  _id: "id6",
};

export const dispositifsWithTranslations = [
  content1,
  content2,
  content3,
  content4,
  content5,
  content6,
];
