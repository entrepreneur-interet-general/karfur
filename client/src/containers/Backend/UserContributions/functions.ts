import { IUserContribution, IDispositif } from "../../../types/interface";
import { FormattedUserContribution } from "./types";

export const formatContributions = (
  userContributions: IUserContribution[],
  userStructureContributions: IDispositif[],
  userStructureName: string | null
): FormattedUserContribution[] => {
  let formattedContribs: FormattedUserContribution[] = [];

  userContributions.forEach((dispositif) => {
    if (["Brouillon", "En attente"].includes(dispositif.status)) {
      return formattedContribs.push({ ...dispositif, responsabilite: "Moi" });
    }
    return formattedContribs.push({
      ...dispositif,
      responsabilite: dispositif.mainSponsor,
    });
  });

  userStructureContributions
    .filter((dispositif) => {
      if (["Brouillon", "SupprimÃ©"].includes(dispositif.status)) return false;

      if (dispositif.status === "En attente") {
        const isDispositifInUserContributions =
          userContributions.filter(
            (userDispo) => userDispo._id === dispositif._id
          ).length > 0;
        if (isDispositifInUserContributions) return false;
      }
      return true;
    })
    .forEach((dispositif) => {
      return formattedContribs.push({
        titreInformatif: dispositif.titreInformatif,
        titreMarque: dispositif.titreMarque,
        typeContenu: dispositif.typeContenu,
        nbMercis: dispositif.nbMercis,
        nbVues: dispositif.nbVues,
        _id: dispositif._id,
        status: dispositif.status,
        responsabilite: userStructureName,
      });
    });

  return formattedContribs;
};
