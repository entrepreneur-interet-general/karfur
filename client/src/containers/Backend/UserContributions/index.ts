export { UserContributionsComponent as UserContributions } from "../UserContributions/UserContributions.component";
