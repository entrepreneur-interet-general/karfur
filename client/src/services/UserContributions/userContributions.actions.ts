import {
  FETCH_USER_CONTRIBUTIONS,
  SET_USER_CONTRIBUTIONS,
  UPDATE_USER_CONTRIBUTIONS,
} from "./userContributions.actionTypes";
import { ActionType, action } from "typesafe-actions";
import { IUserContribution } from "../../types/interface";
import { ObjectId } from "mongodb";

export const fetchUserContributionsActionCreator = () =>
  action(FETCH_USER_CONTRIBUTIONS);

export const setUserContributionsActionCreator = (value: IUserContribution[]) =>
  action(SET_USER_CONTRIBUTIONS, value);

export const updateUserContributionsActionCreator = (value: {
  dispositifId?: ObjectId;
  type: "remove" | "remove-all";
  locale: string;
}) => action(UPDATE_USER_CONTRIBUTIONS, value);

const actions = {
  fetchUserContributionsActionCreator,
  setUserContributionsActionCreator,
  updateUserContributionsActionCreator,
};

export type UserContributionsActions = ActionType<typeof actions>;
