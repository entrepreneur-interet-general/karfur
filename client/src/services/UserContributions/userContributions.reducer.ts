import { createReducer } from "typesafe-actions";
import { UserContributionsActions } from "./userContributions.actions";
import { IUserContribution } from "../../types/interface";

export type UserContributionsState = IUserContribution[];

const initialUserContriubtionsState: UserContributionsState = [];

export const userContributionsReducer = createReducer<
  UserContributionsState,
  UserContributionsActions
>(initialUserContriubtionsState, {
  SET_USER_CONTRIBUTIONS: (_, action) => action.payload,
});
