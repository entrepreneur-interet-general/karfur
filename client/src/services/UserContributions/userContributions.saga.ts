import { SagaIterator } from "redux-saga";
import { logger } from "../../logger";
import { put, call, takeLatest } from "redux-saga/effects";
import {
  startLoading,
  LoadingStatusKey,
  finishLoading,
} from "../LoadingStatus/loadingStatus.actions";
import API from "../../utils/API";
import { FETCH_USER_CONTRIBUTIONS } from "./userContributions.actionTypes";
import { setUserContributionsActionCreator } from "./userContributions.actions";

export function* fetchUserContributions(): SagaIterator {
  try {
    logger.info("[fetchUserContributions] saga");
    yield put(startLoading(LoadingStatusKey.FETCH_USER_CONTRIBUTIONS));
    const data = yield call(API.getUserContributions);
    yield put(setUserContributionsActionCreator(data.data.data));
    yield put(finishLoading(LoadingStatusKey.FETCH_USER_CONTRIBUTIONS));
  } catch (error) {
    logger.error("[fetchUserContributions] saga error", { error });
    yield put(setUserContributionsActionCreator([]));
    yield put(finishLoading(LoadingStatusKey.FETCH_USER_CONTRIBUTIONS));
  }
}

//   export function* updateUserFavorites(
//     action: ReturnType<typeof updateUserFavoritesActionCreator>
//   ): SagaIterator {
//     try {
//       logger.info("[updateUserFavorites] saga", { data: action.payload });
//       yield put(startLoading(LoadingStatusKey.UPDATE_USER_CONTRIBUTIONS));
//       yield call(API.updateUserFavorites, {
//         dispositifId: action.payload.dispositifId,
//         type: action.payload.type,
//       });
//       yield put(fetchUserContributionsActionCreator(action.payload.locale));
//       yield put(finishLoading(LoadingStatusKey.UPDATE_USER_CONTRIBUTIONS));
//     } catch (error) {
//       logger.error("[updateUserFavorites] saga error", { error });
//       yield put(finishLoading(LoadingStatusKey.UPDATE_USER_CONTRIBUTIONS));
//     }
//   }

function* latestActionsSaga() {
  yield takeLatest(FETCH_USER_CONTRIBUTIONS, fetchUserContributions);
  // yield takeLatest(UPDATE_USER_CONTRIBUTIONS, updateUserFavorites);
}

export default latestActionsSaga;
