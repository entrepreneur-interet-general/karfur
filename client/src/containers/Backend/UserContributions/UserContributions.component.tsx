import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchUserContributionsActionCreator } from "../../../services/UserContributions/userContributions.actions";
import { userContributionsSelector } from "../../../services/UserContributions/userContributions.selectors";
import {
  userStructureDisposAssociesSelector,
  userStructureNameSelector,
} from "../../../services/UserStructure/userStructure.selectors";
import { isLoadingSelector } from "../../../services/LoadingStatus/loadingStatus.selectors";
import { LoadingStatusKey } from "../../../services/LoadingStatus/loadingStatus.actions";
import { formatContributions } from "./functions";
import styled from "styled-components";
import { NoContribution } from "./components/NoContribution";
import { FrameModal } from "../../../components/Modals";
import { ContribContainer } from "./components/SubComponents";
import { TitleWithNumber } from "../middleOfficeSharedComponents";
import FButton from "../../../components/FigmaUI/FButton/FButton";
import { NavHashLink } from "react-router-hash-link";
import { UserContribTable } from "./components/UserContribTable";

const MainContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 1;
  margin-top: 42px;
  height: fit-content;
  margin-bottom: 42px;
`;

const TitleContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

const WhiteContainer = styled.div`
  background: #ffffff;
  border-radius: 12px;
  width: 100%;
  padding: 32px;
`;

export const UserContributionsComponent = () => {
  const [showTutoModal, setShowTutoModal] = useState(false);
  const [tutoModalDisplayed, setTutoModalDisplayed] = useState("");
  const toggleTutoModal = () => setShowTutoModal(!showTutoModal);

  const dispatch = useDispatch();

  const userContributions = useSelector(userContributionsSelector);
  const userStructureContributions = useSelector(
    userStructureDisposAssociesSelector
  );
  const userStructureName = useSelector(userStructureNameSelector);
  const isLoadingUserContrib = useSelector(
    isLoadingSelector(LoadingStatusKey.FETCH_USER_CONTRIBUTIONS)
  );
  const isLoadingUserStructureContrib = useSelector(
    isLoadingSelector(LoadingStatusKey.FETCH_USER_STRUCTURE)
  );
  const isLoading = isLoadingUserContrib || isLoadingUserStructureContrib;

  useEffect(() => {
    dispatch(fetchUserContributionsActionCreator());
  }, []);

  const contributions = formatContributions(
    userContributions,
    userStructureContributions,
    userStructureName
  );

  if (isLoading) {
    return <div>Loading</div>;
  }

  if (contributions.length === 0)
    return (
      <MainContainer>
        <NoContribution
          toggleTutoModal={toggleTutoModal}
          setTutoModalDisplayed={setTutoModalDisplayed}
        />
        {showTutoModal && (
          <FrameModal
            show={showTutoModal}
            toggle={toggleTutoModal}
            section={"Mes fiches"}
          />
        )}
      </MainContainer>
    );
  return (
    <MainContainer>
      <ContribContainer>
        <TitleContainer>
          <TitleWithNumber
            amount={contributions.length}
            textPlural="fiches."
            textSingular="fiche."
            textBefore="Vous avez rédigé"
          />
          <div>
            <FButton
              type="tuto"
              name="video-outline"
              className="mr-8"
              onClick={() => {
                setTutoModalDisplayed("Mes fiches");
                toggleTutoModal();
              }}
            >
              Explications
            </FButton>
            <FButton
              tag={NavHashLink}
              to="/comment-contribuer#ecrire"
              type="dark"
              name="file-add-outline"
            >
              Créer une nouvelle fiche
            </FButton>
          </div>
        </TitleContainer>
        <WhiteContainer>
          <UserContribTable
            contributions={contributions}
            toggleTutoModal={toggleTutoModal}
            setTutoModalDisplayed={setTutoModalDisplayed}
          />
        </WhiteContainer>
      </ContribContainer>
      {showTutoModal && (
        <FrameModal
          show={showTutoModal}
          toggle={toggleTutoModal}
          section={tutoModalDisplayed}
        />
      )}
    </MainContainer>
  );
};
