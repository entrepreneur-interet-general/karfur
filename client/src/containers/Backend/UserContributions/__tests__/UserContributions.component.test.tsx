// @ts-nocheck
import { UserContributionsComponent } from "../UserContributions.component";
import { initialMockStore } from "../../../../__fixtures__/reduxStore";
import { wrapWithProvidersAndRender } from "../../../../../jest/lib/wrapWithProvidersAndRender";

describe("userContributions", () => {
  it("should render correctly when loading", () => {
    window.scrollTo = jest.fn();

    const component = wrapWithProvidersAndRender({
      Component: UserContributionsComponent,
      reduxState: {
        ...initialMockStore,
        loadingStatus: { FETCH_USER_CONTRIBUTIONS: { isLoading: true } },
      },
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render correctly when 0 contributions", () => {
    window.scrollTo = jest.fn();
    const component = wrapWithProvidersAndRender({
      Component: UserContributionsComponent,
    });

    expect(component.toJSON()).toMatchSnapshot();
  });
});
