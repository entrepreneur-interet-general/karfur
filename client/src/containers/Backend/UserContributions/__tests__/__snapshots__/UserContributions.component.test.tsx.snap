// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userContributions should render correctly when 0 contributions 1`] = `
<div
  className="sc-fBuWsC gtEXvd"
>
  <div
    className="sc-dnqmqq kCBvsT"
  >
    <div
      className="sc-VigVT iceOHV"
    >
      Vous n'avez pas encore rédigé de fiche.
    </div>
    <div
      className="sc-jTzLTM Xbnql"
    >
      <div
        className="sc-fjdhpX cxxdfG"
      >
        C'est parti ?
      </div>
      <div
        className="sc-jzJRlG jmjfTH"
      >
        Chacun peut apporter sa contribution pour enrichir le site.
      </div>
      <div
        className="sc-jzJRlG jmjfTH"
      >
        Lancez-vous !
      </div>
      <div
        className="sc-cSHVUG iOhvjj"
      >
        <a
          aria-current={null}
          className="figma-btn dark mr-10"
          href="/comment-contribuer#ecrire"
          onClick={[Function]}
        >
          <span
            className="mr-10"
          >
            <i
              data-eva="file-add-outline"
              data-eva-fill="#fff"
              data-eva-height="20px"
              data-eva-width="20px"
            />
          </span>
          Créer une nouvelle fiche
        </a>
        <button
          className="figma-btn tuto ml-10"
          onClick={[Function]}
        >
          <span
            className="mr-10"
          >
            <i
              data-eva="video-outline"
              data-eva-fill="#fff"
              data-eva-height="20px"
              data-eva-width="20px"
            />
          </span>
          Explications
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`userContributions should render correctly when loading 1`] = `
<div
  className="sc-fBuWsC gtEXvd"
>
  <div
    className="sc-dnqmqq kCBvsT"
  >
    <div
      className="sc-bRBYWo gykBnX"
    >
      <div
        className="sc-jhAzac BLwFE"
      >
        Vous avez rédigé
      </div>
      <div
        className="sc-hzDkRC foJWvd"
      >
        ...
      </div>
      <div
        className="sc-jhAzac BLwFE"
      >
        fiche.
      </div>
    </div>
    <span>
      <span
        className="react-loading-skeleton css-1vmnjpn-skeletonStyles-Skeleton"
        style={
          Object {
            "height": 50,
          }
        }
      >
        ‌
      </span>
      <span
        className="react-loading-skeleton css-1vmnjpn-skeletonStyles-Skeleton"
        style={
          Object {
            "height": 50,
          }
        }
      >
        ‌
      </span>
      <span
        className="react-loading-skeleton css-1vmnjpn-skeletonStyles-Skeleton"
        style={
          Object {
            "height": 50,
          }
        }
      >
        ‌
      </span>
    </span>
  </div>
</div>
`;
